packages/ember-simple-auth/addon/services/session.js

```diff
@@ -139,12 +161,61 @@ export default Service.extend(Evented, {
       // the internal session won't be available in route unit tests
       if (session) {
         session.on(event, () => {
+          enableEventsDeprecation = false;
           this.trigger(event, ...arguments);
+          enableEventsDeprecation = true;
         });
       }
     });
   },
 
+  on() {
+    deprecateSessionEvents();
+
+    return this._super(...arguments);
+  },
+
+  one() {
+    deprecateSessionEvents();
+
+    return this._super(...arguments);
+  },
+
+  off() {
+    deprecateSessionEvents();
+
+    return this._super(...arguments);
+  },
+
+  has() {
+    deprecateSessionEvents();
+
+    return this._super(...arguments);
+  },
+
+  trigger() {
+    deprecateSessionEvents();
+
+    return this._super(...arguments);
+  },
+
+  /**
+    This sets the session up for usage in an appliation. In particular, it
+    installs automatic handling of session events like authentication and
+    invalidation.
+
+    __This method must be called as early as possible in the application's
+    lifecycle, specifically in an instance initializer or the application
+    route's constructor.__
+
+    @method setup
+    @public
+  */
+  setup() {
+    this.get('session').on('authenticationSucceeded', () => this.handleAuthentication(Configuration.routeAfterAuthentication));
+    this.get('session').on('invalidationSucceeded', () => this.handleInvalidation(Configuration.rootURL));
+  },
```

Comment:
Why does the internal session communicates with the public session through events? Wondering if internal session can call `handleAuthentication` and `handeInvalidation` methods of public session service directly. This would also remove the need for this `setup()` method, wouldn't it?

Reply:
I think the events between the internal session and the session service are an internal detail that I'd like not to change as part of this. There might be better ways to do this eventually of course. You are right though that we might not need the `setup` method since we're already instantiating the `session` service when the application starts up: https://github.com/simplabs/ember-simple-auth/blob/master/packages/ember-simple-auth/addon/initializers/setup-session-restoration.js#L16

Reply:
I removed the setup method.